cmake/modules/create_python_dist.cmake

```diff
@@ -29,27 +29,29 @@
 ## the package in the build tree. The second argument is the Python module or
 ## Python package that is expected to exist in the current source directory.
 ## Currently, the package/module distinction is made if the second argument is
-## a directory or not. If it is a directory, then it will assume it is a
-## package (and will create an empty __init__.py file in the binary directory
-## if it does not exits). It is up to the developer of the package to ensure
-## that the contents of the directory contain all required modules and the
-## setup.py is correct. It is assumed that the name of the module or package
-## given as the second argument is the name of the package listed in the setup.py
-## file. In either case (if the input is a package or a module), all input files
-## will be processed with 'configure_file' to replace any CMake variables that use
-## the @ symbol
+## a directory or not. If it is a directory, then it must contain an
+## __init__.py file to be treated as a Python package. It is up to the
+## developer of the package to ensure that the contents of the directory
+## contain all required modules and the setup.py is correct. It is assumed that
+## the name of the module or package given as the second argument is the name
+## of the package listed in the setup.py file. In either case (if the input is
+## a package or a module), all input files will be processed with
+## 'configure_file' to replace any CMake variables that use the @ symbol
 function(create_python_dist target mod_or_pkg)
     if(IS_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/${mod_or_pkg})
-        file(GLOB py_src RELATIVE ${CMAKE_CURRENT_SOURCE_DIR}/${mod_or_pkg} ${mod_or_pkg}/*.py)
+        if (NOT EXISTS ${CMAKE_CURRENT_BINARY_DIR}/${mod_or_pkg}/__init__.py)
+            message(SEND_ERROR "${mod_or_pkg} is a directory but missing __init__.py")
+        endif (NOT EXISTS ${CMAKE_CURRENT_BINARY_DIR}/${mod_or_pkg}/__init__.py)
+        file(GLOB_RECURSE py_src
+            RELATIVE
+            ${CMAKE_CURRENT_SOURCE_DIR}/${mod_or_pkg}
+            ${mod_or_pkg}/*.py)
```

Comment:
I didn't believe the regular expression could actually work (${mod_or_pkg}/*.py). It looked as though it ought to match just top level, but it is what CMake expects. Odd choice on their part!

Can you imagine a case where we want to embed non-Python files from a tree? Perhaps some JSON data file, or other configuration? Should we just include everything in the tree and not limit it to Python files? (Perhaps exclude binaries like .pyc.)

Reply:
I have thought about this some and go back and forth. Let me know what you think of these options:
* Current implementation: specify module (`file.py`) or package (`dir/__init__.py`) and have cmake process the directory and call configure_file on each file in the tree. I did this in case any cmake variables need to be replaced. The reason I went this direction is because each file should have a \_\_version\_\_ attribute that is independent of the package version.
Take a look at [PEP 396](https://www.python.org/dev/peps/pep-0396/). 
Specifically, points 3 and 4 of the specification. Now it doesn't require all modules have a version but that modules inside a namespace package should and that they should be in the \_\_version\_\_ attribute. 
* Copy the package directory as is. No cmake variable replacement will be done with the contents. The only file that we call `configure_file` on is `setup.py` as that file should contain the version number
* Use explicit filenames in CMakeLists.txt. I tend to prefer explicit but in this case I think we should treat a package directory as the input leave it up to the Python package developer to ensure contents are valid.
* Process setup.py to get a list of collateral (like data files). We may have to use the setuptools api for this but this may require more work.